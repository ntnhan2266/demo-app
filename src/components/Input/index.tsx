import React from 'react';
import classNames from 'classnames';

interface InputProps {
  type?: 'text' | 'password' | 'email' | 'number';
  label: string;
  placeholder?: string;
  value?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  disabled?: boolean;
  errorMessage?: string;
}

const Input: React.FC<InputProps> = ({
  type = 'text',
  placeholder = '',
  value,
  onChange,
  disabled = false,
  label,
  errorMessage,
}): React.ReactElement => {
  const inputClasses = classNames('border rounded-md px-3 py-2 focus:outline-none transition duration-300 w-full', {
    'border-gray-400': !disabled,
    'bg-gray-200 cursor-not-allowed': disabled,
    'border-red-500': errorMessage,
  });

  const labelClasses = classNames('block text-sm font-semibold mb-1', {
    'text-red-500': errorMessage,
  });

  return (
    <div className='mb-4'>
      <label htmlFor={label} className={labelClasses}>
        {label}
      </label>
      <input
        type={type}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
        className={inputClasses}
        disabled={disabled}
      />
      {errorMessage && <p className='text-red-500 text-xs mt-1'>{errorMessage}</p>}
    </div>
  );
};

export default Input;
