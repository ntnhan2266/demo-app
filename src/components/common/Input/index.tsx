import React from 'react';
import classNames from 'classnames';

interface IProps extends React.InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  errorMessage?: string;
  required?: boolean;
  wrapperClass?: string;
  inputClasses?: string;
}

const Input: React.FC<IProps> = React.forwardRef<HTMLInputElement, IProps>(
  (
    {
      type = 'text',
      wrapperClass = '',
      placeholder = '',
      inputClasses = '',
      value,
      onChange,
      disabled = false,
      required = false,
      label,
      errorMessage,
      id,
      name,
      ...props
    },
    ref,
  ): React.ReactElement => {
    const styleClassnames = classNames(
      `border rounded-md px-3 py-2 focus:outline-none transition duration-300 w-full md:w-96 max-w-full ${inputClasses}`,
      {
        'border-gray-400': !disabled,
        'bg-gray-200 cursor-not-allowed': disabled,
        'border-red-500': errorMessage,
      },
    );

    const labelClasses = classNames('block w-full mb-1 text-xs font-medium text-gray-500 transition-all duration-200 ease-in-out ', {
      'text-red-600': errorMessage,
    });

    return (
      <div className={wrapperClass}>
        {label && (
          <label htmlFor={id || name} className={labelClasses}>
            {label}
            {required && <sup className='text-red-600'>*</sup>}
          </label>
        )}
        <input
          ref={ref}
          id={id}
          name={name}
          type={type}
          placeholder={placeholder}
          value={value}
          onChange={onChange}
          className={styleClassnames}
          disabled={disabled}
          {...props}
        />
        {errorMessage && <p className='text-red-600 text-xs mt-1'>{errorMessage}</p>}
      </div>
    );
  },
);

Input.displayName = 'InputComponent';

export default Input;
