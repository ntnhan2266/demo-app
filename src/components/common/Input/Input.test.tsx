import { render, screen } from '@testing-library/react';
import Input from '@/components/common/Input';
import renderer from 'react-test-renderer';

describe('Input component', () => {
  test('should match the snapshot', () => {
    const tree = renderer.create(<Input />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders input with label and placeholder', () => {
    render(<Input label='Username' placeholder='Enter your username' />);

    const labelElement = screen.getByText('Username');
    const inputElement = screen.getByPlaceholderText('Enter your username');

    expect(labelElement).toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();
  });

  test('renders required indicator in the label', () => {
    render(<Input label='Email' required />);

    const labelElement = screen.getByText('Email');
    const requiredIndicator = screen.getByText('*');

    expect(labelElement).toContainElement(requiredIndicator);
  });
});
