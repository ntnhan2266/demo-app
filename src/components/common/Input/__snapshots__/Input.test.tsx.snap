// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should match the snapshot 1`] = `
<div
  className=""
>
  <input
    className="border rounded-md px-3 py-2 focus:outline-none transition duration-300 w-full md:w-96 max-w-full  border-gray-400"
    disabled={false}
    placeholder=""
    type="text"
  />
</div>
`;
